ksql-common/src/main/java/io/confluent/ksql/properties/with/ConfigMetaData.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.properties.with;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.errorprone.annotations.Immutable;
+import io.confluent.ksql.configdef.UnmodifiableConfigDef;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.kafka.common.config.ConfigDef;
+
+/**
+ * Holds metadata about config defs.
+ */
+@Immutable
+public final class ConfigMetaData {
```

Comment:
The data built here about the config def is used later to implement things in the with clause property classes like:
-  `toString` - which used the _ordered_ list of properties, so we have consistent ordering
-  `toValues` - which uses the list of `Type.SHORT` properties, as these need special handling.
etc.